since I am living in seattle, I decide to do this data science project for Seattle Crime data.
the website for my data source is https://data.seattle.gov/Public-Safety/Crime-Data/4fs7-3vj5
Overall very credible data source provided by seattle local government so
I trust my analysis done on this dataset. Seattle doesn't provide location details on crime events
for privacy reasons

In [2]:
from csv import reader
from pyspark.sql import Row 
from pyspark.sql import SparkSession
from pyspark.sql.types import *
import pandas as pd
import numpy as np
import seaborn as sb
import matplotlib.pyplot as plt
import warnings
import os
os.environ["PYSPARK_PYTHON"] = "python3"

In [3]:
# download crime data from seattle government website. 
import urllib.request
urllib.request.urlretrieve("https://data.seattle.gov/api/views/4fs7-3vj5/rows.csv?accessType=DOWNLOAD&bom=true&format=true&delimiter=,", "/tmp/crime.csv")
dbutils.fs.mv("file:/tmp/crime.csv", "dbfs:/laioffer/spark_hw1/data/seattleCrime.csv")
display(dbutils.fs.ls("dbfs:/laioffer/spark_hw1/data/"))

path name size dbfs:/laioffer/spark_hw1/data/seattleCrime.csv seattleCrime.csv 53602045 dbfs:/laioffer/spark_hw1/data/sf_03_18.csv sf_03_18.csv 550859692

In [4]:
data_path = "dbfs:/laioffer/spark_hw1/data/seattleCrime.csv"

In [5]:
from pyspark.sql import SparkSession
spark = SparkSession \
    .builder \
    .appName("crime analysis") \
    .config("spark.some.config.option", "some-value") \
    .getOrCreate()

df_crime = spark.read.format("csv").option("header", "true").load(data_path)
display(df_crime)
df_crime.createOrReplaceTempView("seattleCrimeTable")

Report Number Occurred Date Occurred Time Reported Date Reported Time Crime Subcategory Primary Offense Description Precinct Sector Beat Neighborhood 20080000465209 12/13/1908 2114 12/13/2008 2114 DUI DUI-LIQUOR EAST G G2 CENTRAL AREA/SQUIRE PARK 20100000202011 06/15/1964 0 06/15/2010 1031 FAMILY OFFENSE-NONVIOLENT CHILD-OTHER WEST Q Q2 QUEEN ANNE 2012000023437 01/01/1973 0 01/25/2012 1048 SEX OFFENSE-OTHER SEXOFF-OTHER NORTH N N2 NORTHGATE 20130000327785 06/01/1974 0 09/09/2013 1117 SEX OFFENSE-OTHER SEXOFF-OTHER UNKNOWN null null UNKNOWN 20160000289384 01/01/1975 0 08/11/2016 1054 SEX OFFENSE-OTHER SEXOFF-OTHER UNKNOWN null null UNKNOWN 1975000079415 12/16/1975 900 12/16/1975 1500 BURGLARY-RESIDENTIAL BURGLARY-FORCE-RES SOUTH R R3 LAKEWOOD/SEWARD PARK 1976000069169 01/01/1976 1 01/31/1976 2359 SEX OFFENSE-OTHER SEXOFF-INDECENT LIBERTIES UNKNOWN null null UNKNOWN 20170000474905 07/01/1976 800 12/27/2017 1519 SEX OFFENSE-OTHER SEXOFF-INDECENT LIBERTIES UNKNOWN null null UNKNOWN 20180000183590 01/01/1977 0 05/22/2018 1328 RAPE SEXOFF-SODOMY UNKNOWN null null UNKNOWN 20090000300201 01/01/1978 1 08/25/2009 1016 SEX OFFENSE-OTHER SEXOFF-OTHER SOUTHWEST W W1 ALKI 1979000049118 01/28/1979 1600 02/09/1979 1430 CAR PROWL THEFT-CARPROWL EAST G G2 CENTRAL AREA/SQUIRE PARK 20100000323749 07/04/1979 0 09/15/2010 1422 SEX OFFENSE-OTHER SEXOFF-OTHER WEST Q Q3 QUEEN ANNE 20100000177050 01/01/1980 952 05/28/2010 952 SEX OFFENSE-OTHER SEXOFF-OTHER UNKNOWN null null UNKNOWN 20120000373012 01/01/1980 0 10/31/2012 513 SEX OFFENSE-OTHER SEXOFF-OTHER SOUTHWEST F F1 HIGH POINT 1981000076456 02/14/1981 2000 02/15/1981 435 BURGLARY-RESIDENTIAL BURGLARY-FORCE-RES SOUTHWEST W W3 ROXHILL/WESTWOOD/ARBOR HEIGHTS 19810000305573 08/22/1981 2029 08/22/1981 2030 HOMICIDE HOMICIDE-PREMEDITATED-WEAPON SOUTH S S2 BRIGHTON/DUNLAP 20150000204962 01/01/1985 1200 06/18/2015 1322 SEX OFFENSE-OTHER SEXOFF-OTHER UNKNOWN null null UNKNOWN 20140000342519 09/13/1985 1300 10/13/2014 923 SEX OFFENSE-OTHER SEXOFF-OTHER EAST G G1 FIRST HILL 20160000454935 01/19/1986 0 12/19/2016 1504 SEX OFFENSE-OTHER SEXOFF-OTHER SOUTH S S2 BRIGHTON/DUNLAP 201300007252 01/01/1987 0 01/07/2013 817 RAPE SEXOFF-SODOMY UNKNOWN null null UNKNOWN 19880000507482 09/29/1988 155 09/29/1988 155 MOTOR VEHICLE THEFT VEH-THEFT-AUTO WEST M M2 SLU/CASCADE 2013000033940 01/01/1989 0 01/30/2013 1425 SEX OFFENSE-OTHER SEXOFF-OTHER SOUTHWEST W W2 ALASKA JUNCTION 20110000415050 01/01/1990 1 12/29/2011 1647 SEX OFFENSE-OTHER SEXOFF-OTHER UNKNOWN null null UNKNOWN 2015000057217 02/19/1990 1223 02/19/2015 1223 FAMILY OFFENSE-NONVIOLENT CHILD-OTHER UNKNOWN null null UNKNOWN 2019000079185 01/01/1991 1 03/04/2019 1934 SEX OFFENSE-OTHER SEXOFF-INDECENT LIBERTIES SOUTHWEST F F3 HIGHLAND PARK 20190000159152 01/01/1991 1 05/04/2019 1304 RAPE RAPE-STRONGARM NORTH L L3 SANDPOINT 20090000267882 01/01/1991 0 07/31/2009 940 FAMILY OFFENSE-NONVIOLENT CHILD-ABUSED-NOFORCE SOUTHWEST W W1 NORTH ADMIRAL 20110000134169 01/01/1991 0 04/27/2011 2244 RAPE SEXOFF-SODOMY SOUTHWEST F F3 SOUTH PARK 20170000144379 01/01/1993 1840 04/25/2017 1840 SEX OFFENSE-OTHER SEXOFF-OTHER EAST G G2 CENTRAL AREA/SQUIRE PARK 20140000330048 01/01/1993 100 10/02/2014 1845 THEFT-ALL OTHER THEFT-OTH EAST E E2 CAPITOL HILL 2008000052323 02/09/1993 2200 02/09/2008 1421 RAPE RAPE-WEAPON SOUTHWEST W W3 ROXHILL/WESTWOOD/ARBOR HEIGHTS 20170000460539 07/09/1993 0 12/14/2017 1938 RAPE RAPE-STRONGARM NORTH L L3 LAKECITY 19930000465323 10/08/1993 2213 10/08/1993 2213 HOMICIDE HOMICIDE-PREMEDITATED-GUN SOUTH R R2 CLAREMONT/RAINIER VISTA 20140000157411 01/01/1994 0 05/20/2014 929 FAMILY OFFENSE-NONVIOLENT CHILD-OTHER EAST G G1 FIRST HILL 19940000264046 06/08/1994 0 06/12/1994 844 THEFT-ALL OTHER THEFT-OTH SOUTHWEST F F1 HIGH POINT 20090000188629 01/01/1995 1 06/01/2009 1546 SEX OFFENSE-OTHER SEXOFF-OTHER EAST G G2 CENTRAL AREA/SQUIRE PARK 20160000393888 01/01/1995 0 10/31/2016 900 RAPE RAPE-STRONGARM NORTH J J3 ROOSEVELT/RAVENNA 20100000428504 01/01/1995 0 12/13/2010 122

look at how just for the first couple of rows the Occured Date and Reported Date have
a lot of differences! Makes me question the data integrity. Maybe someone typed it wrong
but maybe it is true. That'll be interesting to look at later. 
For examle first row says a crime Occured in 1908 but was reported in 2008....

confused about what beat column represents. It turns out it is a code name for certain patrol territory
https://en.wikipedia.org/wiki/Beat_(police)

## 1.counts the number of crimes for different category in seattle

we will do so using SQL commands

In [9]:
%sql select count(*) as Count, `Crime Subcategory` from seattleCrimeTable group by `Crime Subcategory` order by Count desc

Count Crime Subcategory 148263 CAR PROWL 54420 THEFT-ALL OTHER 48638 THEFT-SHOPLIFT 46843 BURGLARY-RESIDENTIAL 43529 MOTOR VEHICLE THEFT 23531 BURGLARY-COMMERCIAL 21438 THEFT-BUILDING 17722 TRESPASS 17381 NARCOTIC 15640 AGGRAVATED ASSAULT 12838 DUI 11882 ROBBERY-STREET 11091 THEFT-BICYCLE 8852 BURGLARY-RESIDENTIAL-SECURE PARKING 6922 AGGRAVATED ASSAULT-DV 6855 FAMILY OFFENSE-NONVIOLENT 6396 SEX OFFENSE-OTHER 4987 WEAPON 4660 ROBBERY-COMMERCIAL 3609 PROSTITUTION 2023 RAPE 1651 LIQUOR LAW VIOLATION 1164 BURGLARY-COMMERCIAL-SECURE PARKING 1071 ARSON 1062 ROBBERY-RESIDENTIAL 283 DISORDERLY CONDUCT 275 HOMICIDE 262 null 195 PORNOGRAPHY 91 LOITERING 17 GAMBLE

looks like the most crime in Seattle area is Car Prowl, which probably makes sense since
people here are rich lol.. They have nice cars...

##2. Count the number of crimes for different regions (Precinct + Neighborhood) and visulize it - histogram

In [12]:
%sql select count(*) as Count, `Precinct` from seattleCrimeTable group by `Precinct` order by Count desc

Count Precinct 168392 NORTH 150614 WEST 77475 EAST 74426 SOUTH 49332 SOUTHWEST 3346 UNKNOWN 6 null

In [13]:
%sql select count(*) as Count, `Neighborhood` from seattleCrimeTable group by `Neighborhood` order by Count desc

Count Neighborhood 48942 DOWNTOWN COMMERCIAL 30820 NORTHGATE 30735 CAPITOL HILL 27402 QUEEN ANNE 23343 SLU/CASCADE 20868 UNIVERSITY 19451 ROOSEVELT/RAVENNA 15611 BALLARD SOUTH 15401 CHINATOWN/INTERNATIONAL DISTRICT 15370 BELLTOWN 13991 FIRST HILL 13563 LAKECITY 12217 CENTRAL AREA/SQUIRE PARK 11405 GREENWOOD 11109 BALLARD NORTH 10649 SANDPOINT 10242 FREMONT 10008 WALLINGFORD 9998 BITTERLAKE 9282 NORTH BEACON HILL 8889 SODO 8817 PIONEER SQUARE 8220 ROXHILL/WESTWOOD/ARBOR HEIGHTS 7876 MAGNOLIA 7089 ALASKA JUNCTION 7070 BRIGHTON/DUNLAP 7029 MOUNT BAKER 6568 MADRONA/LESCHI 6003 GEORGETOWN 5880 RAINIER BEACH 5865 MID BEACON HILL 5633 HIGHLAND PARK 5492 RAINIER VIEW 4953 NORTH ADMIRAL 4675 PHINNEY RIDGE 4471 MORGAN 4140 JUDKINS PARK/NORTH BEACON HILL 4029 SOUTH PARK 3719 MONTLAKE/PORTAGE BAY 3554 NORTH DELRIDGE 3465 EASTLAKE - WEST 3445 HIGH POINT 3433 MILLER PARK 3396 LAKEWOOD/SEWARD PARK 3385 COLUMBIA CITY 3366 UNKNOWN 3194 NEW HOLLY 2632 HILLMAN CITY 2507 CLAREMONT/RAINIER VISTA 2485 ALKI 2375 FAUNTLEROY SW 2251 SOUTH BEACON HILL 1973 SOUTH DELRIDGE 1789 MADISON PARK 1552 GENESEE 860 EASTLAKE - EAST 609 PIGEON POINT 318 COMMERCIAL DUWAMISH 177 COMMERCIAL HARBOR ISLAND

Looks like North area and Downtown + NorthGate have the most crimes going on!
- I live near NorthGate (studying in UW).... Better watch out.

#### Q3: Count the number of crimes happened on "Sunday" at "Seattle downtown".
Note we don't have comprehensive data as in SF crime data. -> we don't have
x, y coordinates. So I am going to do this by filtering in Neighborhood column
(where Neighborhood = 'Downtown Commercial')

was going to use weekday function from sql, but that only deals with "2020-03-01".
Our date format is	"01/01/1991" -> need to use spark function to convert the string 
to date.

In [17]:
from pyspark.sql.functions import *
df_update = df_crime.withColumn("Occurred Date", to_date(col("Occurred Date"), "MM/dd/yyyy"))
df_update.createOrReplaceTempView("seattleCrimeTable")
# now df_update['Occured Date'] has format of year--month--day
# we can use weekday function from sql now!

In [18]:
%sql select count(*) as Count from seattleCrimeTable where  Neighborhood = 'DOWNTOWN COMMERCIAL' and weekday(`Occurred Date`) = 6; 
-- weekday function: 0 means Monday, ... , 6 means Sunday.

Count 6263

So we have 6263 crimes that happened at Seattle Downtown area on a Sunday

#### Q4 question (OLAP)
Analysis the number of crime in each month of 2015, 2016, 2017, 2018. Then, give your insights for the output results. What is the business impact for your result?

In [21]:
%sql select count(*), extract(month from `Occurred Date`) as month from seattleCrimeTable group by month order by month asc
-- first let's see the count of crimes each month without years.

count(1) month 2 null 47133 1 40206 2 44896 3 44469 4 44414 5 42500 6 43602 7 43651 8 43320 9 44789 10 41804 11 42805 12

it seems like the crimes are evenly distributed throughout the months - what I expected

In [23]:
%sql select count(*), extract(year from `Occurred Date`) as year from seattleCrimeTable group by year order by year
-- now look at # crimes each year.

count(1) year 2 null 1 1908 1 1964 1 1973 1 1974 2 1975 2 1976 1 1977 1 1978 2 1979 2 1980 2 1981 2 1985 1 1986 1 1987 1 1988 1 1989 2 1990 4 1991 5 1993 2 1994 6 1995 4 1996 5 1997 20 1998 8 1999 40 2000 53 2001 17 2002 29 2003 43 2004 49 2005 100 2006 627 2007 42793 2008 45056 2009 43353 2010 41298 2011 41007 2012 45551 2013 49322 2014 47693 2015 49220 2016 50334 2017 51302 2018 15624 2019

looks like crimes dramatically increased in 2008. Dropped significantly in 2019, why?
- On the website it says the government starts to collect data in 2008, that is why it peaked there.
- dropped in 2019 probably because this dataset is old, not fully up to date: last updated: August 22, 2019 -> makes sense now why 2019 has less crimes..

In [25]:
%sql select count(*), month(`Occurred Date`) as month from seattleCrimeTable where year(`Occurred Date`) = 2015 group by month order by month
-- # crime each month in 2015

count(1) month 3913 1 3476 2 3851 3 3771 4 4041 5 4002 6 4007 7 4036 8 4010 9 4242 10 4122 11 4222 12

In [26]:
%sql select count(*), month(`Occurred Date`) as month from seattleCrimeTable where year(`Occurred Date`) = 2016 group by month order by month
-- # crime each month in 2016

count(1) month 4400 1 4050 2 3898 3 3909 4 4313 5 3926 6 4201 7 3969 8 3940 9 4204 10 4221 11 4189 12

In [27]:
%sql select count(*), month(`Occurred Date`) as month from seattleCrimeTable where year(`Occurred Date`) = 2017 group by month order by month
-- # crime each month in 2017

count(1) month 4208 1 3552 2 4351 3 4029 4 4223 5 4130 6 4484 7 4552 8 4187 9 4422 10 4028 11 4168 12

In [28]:
%sql select count(*), month(`Occurred Date`) as month from seattleCrimeTable where year(`Occurred Date`) = 2018 group by month order by month
-- # crime each month in 2018

count(1) month 4196 1 3891 2 4197 3 4116 4 4441 5 4376 6 4452 7 4410 8 4328 9 4321 10 4171 11 4403 12

In [29]:
%sql select count(*), month(`Occurred Date`) as month from seattleCrimeTable group by month order by month
-- all time count of crimes vs month

count(1) month 2 null 47133 1 40206 2 44896 3 44469 4 44414 5 42500 6 43602 7 43651 8 43320 9 44789 10 41804 11 42805 12

- Overall the crimes are evenly distributed throughout each month in the year of 2015, 2016, 2017, and 2018.
- However there sometimes is noticeably fewer crimes in February, why is that?
- Business impact: not so obvious. 
  - Distribute police force evenly throughout each year?
  - Maybe since February has less crimes, increase the rent for housing in that month?

#### Q5 question (OLAP)
Analysis the number of crime w.r.t the hour in certian day like 2015/12/15, 2016/12/15, 2017/12/15. Then, give your travel suggestion to visit Seattle.

In [32]:
display(df_crime)

Report Number Occurred Date Occurred Time Reported Date Reported Time Crime Subcategory Primary Offense Description Precinct Sector Beat Neighborhood 20080000465209 12/13/1908 2114 12/13/2008 2114 DUI DUI-LIQUOR EAST G G2 CENTRAL AREA/SQUIRE PARK 20100000202011 06/15/1964 0 06/15/2010 1031 FAMILY OFFENSE-NONVIOLENT CHILD-OTHER WEST Q Q2 QUEEN ANNE 2012000023437 01/01/1973 0 01/25/2012 1048 SEX OFFENSE-OTHER SEXOFF-OTHER NORTH N N2 NORTHGATE 20130000327785 06/01/1974 0 09/09/2013 1117 SEX OFFENSE-OTHER SEXOFF-OTHER UNKNOWN null null UNKNOWN 20160000289384 01/01/1975 0 08/11/2016 1054 SEX OFFENSE-OTHER SEXOFF-OTHER UNKNOWN null null UNKNOWN 1975000079415 12/16/1975 900 12/16/1975 1500 BURGLARY-RESIDENTIAL BURGLARY-FORCE-RES SOUTH R R3 LAKEWOOD/SEWARD PARK 1976000069169 01/01/1976 1 01/31/1976 2359 SEX OFFENSE-OTHER SEXOFF-INDECENT LIBERTIES UNKNOWN null null UNKNOWN 20170000474905 07/01/1976 800 12/27/2017 1519 SEX OFFENSE-OTHER SEXOFF-INDECENT LIBERTIES UNKNOWN null null UNKNOWN 20180000183590 01/01/1977 0 05/22/2018 1328 RAPE SEXOFF-SODOMY UNKNOWN null null UNKNOWN 20090000300201 01/01/1978 1 08/25/2009 1016 SEX OFFENSE-OTHER SEXOFF-OTHER SOUTHWEST W W1 ALKI 1979000049118 01/28/1979 1600 02/09/1979 1430 CAR PROWL THEFT-CARPROWL EAST G G2 CENTRAL AREA/SQUIRE PARK 20100000323749 07/04/1979 0 09/15/2010 1422 SEX OFFENSE-OTHER SEXOFF-OTHER WEST Q Q3 QUEEN ANNE 20100000177050 01/01/1980 952 05/28/2010 952 SEX OFFENSE-OTHER SEXOFF-OTHER UNKNOWN null null UNKNOWN 20120000373012 01/01/1980 0 10/31/2012 513 SEX OFFENSE-OTHER SEXOFF-OTHER SOUTHWEST F F1 HIGH POINT 1981000076456 02/14/1981 2000 02/15/1981 435 BURGLARY-RESIDENTIAL BURGLARY-FORCE-RES SOUTHWEST W W3 ROXHILL/WESTWOOD/ARBOR HEIGHTS 19810000305573 08/22/1981 2029 08/22/1981 2030 HOMICIDE HOMICIDE-PREMEDITATED-WEAPON SOUTH S S2 BRIGHTON/DUNLAP 20150000204962 01/01/1985 1200 06/18/2015 1322 SEX OFFENSE-OTHER SEXOFF-OTHER UNKNOWN null null UNKNOWN 20140000342519 09/13/1985 1300 10/13/2014 923 SEX OFFENSE-OTHER SEXOFF-OTHER EAST G G1 FIRST HILL 20160000454935 01/19/1986 0 12/19/2016 1504 SEX OFFENSE-OTHER SEXOFF-OTHER SOUTH S S2 BRIGHTON/DUNLAP 201300007252 01/01/1987 0 01/07/2013 817 RAPE SEXOFF-SODOMY UNKNOWN null null UNKNOWN 19880000507482 09/29/1988 155 09/29/1988 155 MOTOR VEHICLE THEFT VEH-THEFT-AUTO WEST M M2 SLU/CASCADE 2013000033940 01/01/1989 0 01/30/2013 1425 SEX OFFENSE-OTHER SEXOFF-OTHER SOUTHWEST W W2 ALASKA JUNCTION 20110000415050 01/01/1990 1 12/29/2011 1647 SEX OFFENSE-OTHER SEXOFF-OTHER UNKNOWN null null UNKNOWN 2015000057217 02/19/1990 1223 02/19/2015 1223 FAMILY OFFENSE-NONVIOLENT CHILD-OTHER UNKNOWN null null UNKNOWN 2019000079185 01/01/1991 1 03/04/2019 1934 SEX OFFENSE-OTHER SEXOFF-INDECENT LIBERTIES SOUTHWEST F F3 HIGHLAND PARK 20190000159152 01/01/1991 1 05/04/2019 1304 RAPE RAPE-STRONGARM NORTH L L3 SANDPOINT 20090000267882 01/01/1991 0 07/31/2009 940 FAMILY OFFENSE-NONVIOLENT CHILD-ABUSED-NOFORCE SOUTHWEST W W1 NORTH ADMIRAL 20110000134169 01/01/1991 0 04/27/2011 2244 RAPE SEXOFF-SODOMY SOUTHWEST F F3 SOUTH PARK 20170000144379 01/01/1993 1840 04/25/2017 1840 SEX OFFENSE-OTHER SEXOFF-OTHER EAST G G2 CENTRAL AREA/SQUIRE PARK 20140000330048 01/01/1993 100 10/02/2014 1845 THEFT-ALL OTHER THEFT-OTH EAST E E2 CAPITOL HILL 2008000052323 02/09/1993 2200 02/09/2008 1421 RAPE RAPE-WEAPON SOUTHWEST W W3 ROXHILL/WESTWOOD/ARBOR HEIGHTS 20170000460539 07/09/1993 0 12/14/2017 1938 RAPE RAPE-STRONGARM NORTH L L3 LAKECITY 19930000465323 10/08/1993 2213 10/08/1993 2213 HOMICIDE HOMICIDE-PREMEDITATED-GUN SOUTH R R2 CLAREMONT/RAINIER VISTA 20140000157411 01/01/1994 0 05/20/2014 929 FAMILY OFFENSE-NONVIOLENT CHILD-OTHER EAST G G1 FIRST HILL 19940000264046 06/08/1994 0 06/12/1994 844 THEFT-ALL OTHER THEFT-OTH SOUTHWEST F F1 HIGH POINT 20090000188629 01/01/1995 1 06/01/2009 1546 SEX OFFENSE-OTHER SEXOFF-OTHER EAST G G2 CENTRAL AREA/SQUIRE PARK 20160000393888 01/01/1995 0 10/31/2016 900 RAPE RAPE-STRONGARM NORTH J J3 ROOSEVELT/RAVENNA 20100000428504 01/01/1995 0 12/13/2010 122

- our occurred time column has the format of 2114 to denote 21:14, and 920 to denote 9:20. and there are 0, 1, assuming 0 mean 0:00 in the morning and 1 means 1:00 in the morning. (we really don't know since data source didn't describe what these means..)
- Need to deal with that somehow.

In [34]:
%sql select count(*) from seattleCrimeTable where year(`Occurred Date`) = 2015 and month(`Occurred Date`) == 12 and day(`Occurred Date`) == 15
-- see how many crimes on 2015/12/15

count(1) 154

In [35]:
def extractHourFromWeirdFormat(string):
  if not string:
    return None
  if len(string) == 4:
     # 1120 -> 11
    return int(string[:2])
  elif len(string) == 3:
    # 920 -> 9
    return int(string[:1])
  elif len(string) == 1:
    return int(string)
  else:
    return None

In [36]:
from pyspark.sql.functions import col, udf
from pyspark.sql.functions import expr
from pyspark.sql.types import IntegerType

# make our own udf for extracting hour from the given formatted string
hour_func =  udf(lambda string: extractHourFromWeirdFormat(string), IntegerType())

In [37]:
df_update = df_update.withColumn('hour', hour_func(col('Occurred Time')))

In [38]:
display(df_update)

Report Number Occurred Date Occurred Time Reported Date Reported Time Crime Subcategory Primary Offense Description Precinct Sector Beat Neighborhood hour 20080000465209 1908-12-13 2114 12/13/2008 2114 DUI DUI-LIQUOR EAST G G2 CENTRAL AREA/SQUIRE PARK 21 20100000202011 1964-06-15 0 06/15/2010 1031 FAMILY OFFENSE-NONVIOLENT CHILD-OTHER WEST Q Q2 QUEEN ANNE 0 2012000023437 1973-01-01 0 01/25/2012 1048 SEX OFFENSE-OTHER SEXOFF-OTHER NORTH N N2 NORTHGATE 0 20130000327785 1974-06-01 0 09/09/2013 1117 SEX OFFENSE-OTHER SEXOFF-OTHER UNKNOWN null null UNKNOWN 0 20160000289384 1975-01-01 0 08/11/2016 1054 SEX OFFENSE-OTHER SEXOFF-OTHER UNKNOWN null null UNKNOWN 0 1975000079415 1975-12-16 900 12/16/1975 1500 BURGLARY-RESIDENTIAL BURGLARY-FORCE-RES SOUTH R R3 LAKEWOOD/SEWARD PARK 9 1976000069169 1976-01-01 1 01/31/1976 2359 SEX OFFENSE-OTHER SEXOFF-INDECENT LIBERTIES UNKNOWN null null UNKNOWN 1 20170000474905 1976-07-01 800 12/27/2017 1519 SEX OFFENSE-OTHER SEXOFF-INDECENT LIBERTIES UNKNOWN null null UNKNOWN 8 20180000183590 1977-01-01 0 05/22/2018 1328 RAPE SEXOFF-SODOMY UNKNOWN null null UNKNOWN 0 20090000300201 1978-01-01 1 08/25/2009 1016 SEX OFFENSE-OTHER SEXOFF-OTHER SOUTHWEST W W1 ALKI 1 1979000049118 1979-01-28 1600 02/09/1979 1430 CAR PROWL THEFT-CARPROWL EAST G G2 CENTRAL AREA/SQUIRE PARK 16 20100000323749 1979-07-04 0 09/15/2010 1422 SEX OFFENSE-OTHER SEXOFF-OTHER WEST Q Q3 QUEEN ANNE 0 20100000177050 1980-01-01 952 05/28/2010 952 SEX OFFENSE-OTHER SEXOFF-OTHER UNKNOWN null null UNKNOWN 9 20120000373012 1980-01-01 0 10/31/2012 513 SEX OFFENSE-OTHER SEXOFF-OTHER SOUTHWEST F F1 HIGH POINT 0 1981000076456 1981-02-14 2000 02/15/1981 435 BURGLARY-RESIDENTIAL BURGLARY-FORCE-RES SOUTHWEST W W3 ROXHILL/WESTWOOD/ARBOR HEIGHTS 20 19810000305573 1981-08-22 2029 08/22/1981 2030 HOMICIDE HOMICIDE-PREMEDITATED-WEAPON SOUTH S S2 BRIGHTON/DUNLAP 20 20150000204962 1985-01-01 1200 06/18/2015 1322 SEX OFFENSE-OTHER SEXOFF-OTHER UNKNOWN null null UNKNOWN 12 20140000342519 1985-09-13 1300 10/13/2014 923 SEX OFFENSE-OTHER SEXOFF-OTHER EAST G G1 FIRST HILL 13 20160000454935 1986-01-19 0 12/19/2016 1504 SEX OFFENSE-OTHER SEXOFF-OTHER SOUTH S S2 BRIGHTON/DUNLAP 0 201300007252 1987-01-01 0 01/07/2013 817 RAPE SEXOFF-SODOMY UNKNOWN null null UNKNOWN 0 19880000507482 1988-09-29 155 09/29/1988 155 MOTOR VEHICLE THEFT VEH-THEFT-AUTO WEST M M2 SLU/CASCADE 1 2013000033940 1989-01-01 0 01/30/2013 1425 SEX OFFENSE-OTHER SEXOFF-OTHER SOUTHWEST W W2 ALASKA JUNCTION 0 20110000415050 1990-01-01 1 12/29/2011 1647 SEX OFFENSE-OTHER SEXOFF-OTHER UNKNOWN null null UNKNOWN 1 2015000057217 1990-02-19 1223 02/19/2015 1223 FAMILY OFFENSE-NONVIOLENT CHILD-OTHER UNKNOWN null null UNKNOWN 12 2019000079185 1991-01-01 1 03/04/2019 1934 SEX OFFENSE-OTHER SEXOFF-INDECENT LIBERTIES SOUTHWEST F F3 HIGHLAND PARK 1 20190000159152 1991-01-01 1 05/04/2019 1304 RAPE RAPE-STRONGARM NORTH L L3 SANDPOINT 1 20090000267882 1991-01-01 0 07/31/2009 940 FAMILY OFFENSE-NONVIOLENT CHILD-ABUSED-NOFORCE SOUTHWEST W W1 NORTH ADMIRAL 0 20110000134169 1991-01-01 0 04/27/2011 2244 RAPE SEXOFF-SODOMY SOUTHWEST F F3 SOUTH PARK 0 20170000144379 1993-01-01 1840 04/25/2017 1840 SEX OFFENSE-OTHER SEXOFF-OTHER EAST G G2 CENTRAL AREA/SQUIRE PARK 18 20140000330048 1993-01-01 100 10/02/2014 1845 THEFT-ALL OTHER THEFT-OTH EAST E E2 CAPITOL HILL 1 2008000052323 1993-02-09 2200 02/09/2008 1421 RAPE RAPE-WEAPON SOUTHWEST W W3 ROXHILL/WESTWOOD/ARBOR HEIGHTS 22 20170000460539 1993-07-09 0 12/14/2017 1938 RAPE RAPE-STRONGARM NORTH L L3 LAKECITY 0 19930000465323 1993-10-08 2213 10/08/1993 2213 HOMICIDE HOMICIDE-PREMEDITATED-GUN SOUTH R R2 CLAREMONT/RAINIER VISTA 22 20140000157411 1994-01-01 0 05/20/2014 929 FAMILY OFFENSE-NONVIOLENT CHILD-OTHER EAST G G1 FIRST HILL 0 19940000264046 1994-06-08 0 06/12/1994 844 THEFT-ALL OTHER THEFT-OTH SOUTHWEST F F1 HIGH POINT 0 20090000188629 1995-01-01 1 06/01/2009 1546 SEX OFFENSE-OTHER SEXOFF-OTHER EAST G G2 CENTRAL AREA/SQUIRE PARK 1 20160000393888 1995-01-01 0 10/31/2016 900 RAPE

it seems like we are able to extract hour information from Occurred Time column successfully.

In [40]:
df_update.createOrReplaceTempView("seattleCrimeTable") # create a new updated SQL table

In [41]:
%sql select count(*), hour from seattleCrimeTable where year(`Occurred Date`) == 2015 and month(`Occurred Date`) == 12 and day(`Occurred Date`) == 15 group by hour order by hour asc
-- crime count w.r.t hour on 2015/12/15
-- ignore first bar - it is of type null

count(1) hour 4 null 3 0 7 1 3 2 1 3 3 5 2 6 4 7 9 8 6 9 6 10 5 11 7 12 4 13 12 14 8 15 9 16 4 17 16 18 9 19 8 20 8 21 10 22 6 23

In [42]:
%sql select count(*), hour from seattleCrimeTable where year(`Occurred Date`) == 2016 and month(`Occurred Date`) == 12 and day(`Occurred Date`) == 15 group by hour order by hour asc
-- crime count w.r.t hour on 2016/12/15
-- ignore first bar - it is of type null

count(1) hour 2 null 9 0 4 1 1 2 1 3 2 4 2 5 5 6 2 7 2 8 9 9 8 10 9 11 7 12 12 13 10 14 15 15 7 16 12 17 10 18 6 19 11 20 10 21 8 22 1 23

In [43]:
%sql select count(*), hour from seattleCrimeTable where year(`Occurred Date`) == 2017 and month(`Occurred Date`) == 12 and day(`Occurred Date`) == 15 group by hour order by hour asc
-- crime count w.r.t hour on 2017/12/15
-- ignore first bar - it is of type null

count(1) hour 3 null 6 0 2 1 4 2 2 3 2 4 4 6 6 7 1 8 4 9 8 10 6 11 9 12 9 13 11 14 9 15 6 16 12 17 13 18 6 19 4 20 11 21 7 22 5 23

In [44]:
%sql select count(*), hour from seattleCrimeTable group by hour order by hour asc
-- all time crime count w.r.t hour
-- ignore first bar - it is of type null

count(1) hour 11409 null 13421 0 23462 1 14417 2 9841 3 7922 4 7500 5 8636 6 12018 7 17204 8 17108 9 18469 10 19172 11 29296 12 22838 13 23984 14 27200 15 27318 16 31493 17 33495 18 29930 19 30991 20 29917 21 30251 22 26299 23

all three crime number vs hour distributions share similar qualitative behavior: 
- in the morning between 2:00 - 11:00 there are typically less crimes. 
- Crimes happen a lot during from 12:00 - 23:00
- on 2017/12/15 and 2015/12/15 the crimes peaked at 18:00
- all time: 12:00, 18:00, 2:00 seem like the time during day when crimes happen the most

travel suggestion:
- travel in the morning and don't travel late at night
- don't come to Seattle at 6:00 pm.. peaked crime time
- there is almost no safe time... 2:00AM-7:00AM seems the safest

#### Q6 question (OLAP)
(1) Step1: Find out the top-3 danger disrict  
(2) Step2: find out the crime event w.r.t category and time (hour) from the result of step 1  
(3) give your advice to distribute the police based on your analysis results.

In [47]:
%sql select count(*) as Count, `Neighborhood` from seattleCrimeTable group by `Neighborhood` order by Count desc limit 3
-- (1) using neighborhood column to see the top-3 danger district.

Count Neighborhood 48942 DOWNTOWN COMMERCIAL 30820 NORTHGATE 30735 CAPITOL HILL

the top 3 dangerous district for crime in seattle is 1. downtown, 2. northgate, 3. capitol hill

In [49]:
%sql select count(*) as Count, `Crime Subcategory` from seattleCrimeTable where Neighborhood == 'DOWNTOWN COMMERCIAL' group by `Crime Subcategory` order by Count desc
-- downtown w.r.t crime category

Count Crime Subcategory 12295 CAR PROWL 11160 THEFT-SHOPLIFT 5056 THEFT-ALL OTHER 4334 NARCOTIC 2831 THEFT-BUILDING 2210 TRESPASS 1802 BURGLARY-COMMERCIAL 1751 AGGRAVATED ASSAULT 1589 ROBBERY-STREET 866 MOTOR VEHICLE THEFT 793 THEFT-BICYCLE 681 ROBBERY-COMMERCIAL 610 WEAPON 601 SEX OFFENSE-OTHER 601 DUI 371 FAMILY OFFENSE-NONVIOLENT 217 BURGLARY-RESIDENTIAL 201 AGGRAVATED ASSAULT-DV 198 LIQUOR LAW VIOLATION 139 RAPE 129 BURGLARY-RESIDENTIAL-SECURE PARKING 128 BURGLARY-COMMERCIAL-SECURE PARKING 124 PROSTITUTION 80 DISORDERLY CONDUCT 57 PORNOGRAPHY 43 ROBBERY-RESIDENTIAL 36 ARSON 16 HOMICIDE 13 null 7 LOITERING 3 GAMBLE

In [50]:
%sql select count(*) as Count, hour from seattleCrimeTable where Neighborhood == 'DOWNTOWN COMMERCIAL' group by hour order by hour 
-- downtown w.r.t hour

Count hour 809 null 412 0 1105 1 837 2 531 3 457 4 492 5 620 6 992 7 1426 8 1736 9 2110 10 2541 11 3236 12 3126 13 3285 14 3528 15 3344 16 3548 17 3374 18 3057 19 2703 20 2246 21 1898 22 1529 23

In [51]:
%sql select count(*) as Count, `Crime Subcategory` from seattleCrimeTable where Neighborhood == 'NORTHGATE' group by `Crime Subcategory` order by Count desc
-- northgate w.r.t crime category

Count Crime Subcategory 7123 CAR PROWL 5224 THEFT-SHOPLIFT 3026 THEFT-ALL OTHER 2787 MOTOR VEHICLE THEFT 2510 BURGLARY-RESIDENTIAL 1456 PROSTITUTION 1347 THEFT-BUILDING 1205 BURGLARY-COMMERCIAL 1041 TRESPASS 653 AGGRAVATED ASSAULT 627 NARCOTIC 574 DUI 557 ROBBERY-STREET 500 THEFT-BICYCLE 409 ROBBERY-COMMERCIAL 382 FAMILY OFFENSE-NONVIOLENT 359 AGGRAVATED ASSAULT-DV 288 SEX OFFENSE-OTHER 226 BURGLARY-RESIDENTIAL-SECURE PARKING 199 WEAPON 106 RAPE 63 ROBBERY-RESIDENTIAL 58 ARSON 45 LIQUOR LAW VIOLATION 19 BURGLARY-COMMERCIAL-SECURE PARKING 12 null 9 HOMICIDE 7 DISORDERLY CONDUCT 7 PORNOGRAPHY 1 LOITERING

In [52]:
%sql select count(*) as Count, hour from seattleCrimeTable where Neighborhood == 'NORTHGATE' group by hour order by hour
-- northgate w.r.t hour

Count hour 510 null 645 0 1006 1 614 2 470 3 321 4 373 5 519 6 701 7 949 8 903 9 956 10 1096 11 1827 12 1398 13 1658 14 1930 15 2042 16 2162 17 2115 18 1910 19 2029 20 1849 21 1633 22 1204 23

In [53]:
%sql select count(*) as Count, `Crime Subcategory` from seattleCrimeTable where Neighborhood == 'CAPITOL HILL' group by `Crime Subcategory` order by Count desc
-- capitol hill w.r.t crime category

Count Crime Subcategory 6934 CAR PROWL 3527 THEFT-ALL OTHER 3096 THEFT-SHOPLIFT 2108 MOTOR VEHICLE THEFT 1890 THEFT-BUILDING 1844 TRESPASS 1744 BURGLARY-RESIDENTIAL 1222 AGGRAVATED ASSAULT 1207 NARCOTIC 1198 BURGLARY-COMMERCIAL 1135 BURGLARY-RESIDENTIAL-SECURE PARKING 951 DUI 891 THEFT-BICYCLE 826 ROBBERY-STREET 406 SEX OFFENSE-OTHER 394 WEAPON 331 ROBBERY-COMMERCIAL 317 AGGRAVATED ASSAULT-DV 183 FAMILY OFFENSE-NONVIOLENT 165 LIQUOR LAW VIOLATION 110 RAPE 84 BURGLARY-COMMERCIAL-SECURE PARKING 56 ARSON 43 ROBBERY-RESIDENTIAL 28 DISORDERLY CONDUCT 13 PORNOGRAPHY 11 HOMICIDE 10 null 7 PROSTITUTION 4 LOITERING

In [54]:
%sql select count(*) as Count, hour from seattleCrimeTable where Neighborhood == 'CAPITOL HILL' group by hour order by hour
-- capitol hill w.r.t hour

Count hour 1078 null 944 0 2029 1 1484 2 784 3 493 4 471 5 504 6 574 7 773 8 842 9 885 10 909 11 1453 12 1129 13 1151 14 1334 15 1340 16 1624 17 1821 18 1672 19 1838 20 1786 21 1836 22 1981 23

- for capitol hill, we should distribute police force to patrol from 1:00 AM to 2:00 AM + 18:00 PM to 23:00 PM
- for northgate and downtown, distribute police patrol from 12:00 PM to 20:00 PM
- for all three places Car prowl and theft are by far the most common crimes, so police should focus on catching those crimes.
- if need to choose which area to send police, should probably send it to downtown because that is top one crime area.

#### Q7 question (OLAP)
For different category of crime, find the percentage of resolution. Based on the output, give your hints to adjust the policy.

sadly, we don't have resolution column in this data set.. so..😞

### Conclusion. 
- number of crimes has been increasing somewhat steadily from 2008 to 2018, with 2008 having 42,000 crimes in 2008 and 51,000 crimes in 2018. -> maybe the local government should increase police force. Social scientists can help find out why crime has been increasing. - maybe people are poor? Or they lack education? That is another research area
- car prowl and theft are the most common types of crimes committed in seattle over all time. -> citizens better lock their doors.  for preventing shotlifting maybe install cameras in shops. Car insurance companies might need to take high car prowling rate into consideration! Citizens should protect their cars better. Police should focus on catching these crimes. Also 
- North area has the most crimes - send more police to North area of Seattle
- Most of the crimes happen in the afternoon of late night - distribute police to go patrol at these times.
- I really like this project because I used Spark SQL to analyzed big data with 524,000 rows. One important thing I learned is how to make custom udf function to take out important pieces information I want out of raw data.
- also another thing I learned is the need to look at data format - data in the original download link is spearated by ; , and i tried to break it down by , so I failed.. Took me 1 hour to realize the delimeter is wrong, and I downloaded the data with comma being delimeter and all is good
- this kind of work is important to me because analyzing crime data can help the city better plan future police force and the insights can help policy makers to make future policies.